### PR TITLE
Enh/presmoothing

### PR DESCRIPTION
**Pull request by madisoth**
## Changes proposed in this pull request

Adds new option `--presmoothing <fwhm>` to optionally apply spatial smoothing to BOLD / CIFTI timeseries prior to postprocessing. (fMRIPrep/NiBabies do not provide an option to smooth their preprocessed outputs, so this `presmoothing` option is intended as a workaround.)  

If enabled, `desc-residual_bold` outputs will have the specified presmoothing, and `desc-residual_smooth_bold` output will have the specified presmoothing **in addition to** the smoothing specified by `--smoothing` (default 6 mm FWHM).

- Presmoothing is done using a Gaussian kernel of user-specified width (in mm FWHM), the same as `--smoothing`.

- Presmoothing is not enabled by default

- Even if presmoothing is enabled, executive summary "raw" / "pre-regression" figures and measures will still be based on the original, unsmoothed BOLD. 

- added info on presmoothing option in api / generalworkflow docs

### CODE REVIEW

**Comment by kahinimehta** [comment]
We might need to update this before we can merge into main

File: xcp_d/workflow/bold.py [hunk 1]
```diff
@@ -406,6 +417,23 @@ def init_boldpostprocess_wf(lower_bpf,
         n_procs=omp_nthreads,
     )
 
+    # if presmoothing is enabled, use the presmoothed bold file as input to
+    # further preprocessing
+    # else, use the bold file from inputnode
+    bolddatanode = pe.Node(niu.IdentityInterface(
+        fields=['bold_file']),
+        name='bolddatanode')
```
Comment (tsalo) [inline xcp_d/workflow/bold.py:425]: I'd recommend separating the words in the variable name with underscores (e.g., `bold_data_node`).

File: xcp_d/workflow/bold.py [hunk 2]
```diff
@@ -294,6 +298,13 @@ def init_boldpostprocess_wf(lower_bpf,
         mem_gb=mem_gbx['timeseries'],
         omp_nthreads=omp_nthreads)
 
+    presmoothing_wf = init_pre_smoothing(
+        mem_gb=mem_gbx['timeseries'],
+        presmoothing=presmoothing,
+        cifti=False,
+        name="presmoothing_wf",
+        omp_nthreads=omp_nthreads)
```
[the remark this answers is not in the document]
Comment (madisoth) [reply xcp_d/workflow/bold.py]: that's a good idea, would make it consistent with rm_dummytime and despike3d being in their relevant if-statements

File: docs/generalworkflow.rst [hunk 3]
```diff
@@ -15,115 +15,118 @@ both of which require their respective `input-type` flags ( `dcan`` or `hcp` ).
 Processing Steps
 ----------------
 
-0.  Data is read in.
+#.  Data is read in.
     See :ref:`usage_inputs` for information on input dataset structures.
+    
+#. BOLD pre-smoothing [Optional]: Using the option ``--presmoothing``, a Gaussian spatial smoothing kernel of user-specified width (mm FWHM) may be applied to the input BOLD timeseries data before further postprocessing. If enabled, ``desc-residual_bold`` outputs will have the specified presmoothing applied, and ``desc-residual_smooth_bold`` outputs will have the specified presmoothing, plus smoothing of the postprocessed residuals specified by ``--smoothing`` (default 6 mm FWHM).
```
Comment (tsalo) [inline docs/generalworkflow.rst:21]:
```suggestion
#.  BOLD pre-smoothing [Optional]:
     Using the option ``--presmoothing``, a Gaussian spatial smoothing kernel of user-specified width (mm FWHM) may be applied to the input BOLD timeseries data before further postprocessing.
     If enabled, ``desc-residual_bold`` outputs will have the specified presmoothing applied, and ``desc-residual_smooth_bold`` outputs will have the specified presmoothing, plus smoothing of the postprocessed residuals specified by ``--smoothing`` (default 6 mm FWHM).  
```

Just for readability on the dev side of things.

File: docs/generalworkflow.rst [hunk 4]
```diff
@@ -133,43 +136,43 @@ Processing Steps
...
     For each BOLD data, the resting-state derivatives are computed.
     These includes regional homogeneity (ReHo) and amplitude of low-frequency fluctuation (ALFF).
 
-9.  Residual BOLD and resting-state derivatives smoothing:
+#.  Residual BOLD and resting-state derivatives smoothing:
     A smoothing kernel of 6mm is implemented as default for smoothing residual BOLD, ReHo and ALFF.
     Kernel size can be modified in the command line arguments via the ``--smoothing`` flag.
 
-10. Quality control. The quality control (QC) in ``XCP-D`` estimates the quality of BOLD data before and after regression and also estimates BOLD-T1w coregistration and BOLD-Template normalization qualites.
+#. Quality control. The quality control (QC) in ``XCP-D`` estimates the quality of BOLD data before and after regression and also estimates BOLD-T1w coregistration and BOLD-Template normalization qualites.
```
Comment (tsalo) [inline docs/generalworkflow.rst:175]:
```suggestion
#.  Quality control. The quality control (QC) in ``XCP-D`` estimates the quality of BOLD data before and after regression and also estimates BOLD-T1w coregistration and BOLD-Template normalization qualites.
```

File: xcp_d/utils/concantenation.py [hunk 5]
```diff
@@ -221,10 +223,24 @@ def concatenate_nifti(subid, fmridir, outputdir, ses=None, work_dir=None):
...
+            resid = res.split('task-')[1].partition('_')[-1]
+            # print(resid)
+            for j in datafile:
+                fileid = res.split('task-')[0] + resid.partition('_desc')[0]
+                outfile = fileid + j
+                filex = glob.glob(res.split('task-')[0] + '*task*' + j)
+                if j.endswith('framewisedisplacement_bold.tsv'):
+                    name = f"{fileid}{j.split('.')[0]}-DCAN.hdf5"
+                    make_dcan_df(filex, name)
+        if len(resbold_unsmoothed_only) > 1:
```
Comment (tsalo) [inline xcp_d/utils/concantenation.py:243]:
```suggestion
                    make_dcan_df(filex, name)

        if len(resbold_unsmoothed_only) > 1:
```

Could you include empty lines between if statements, for readability?

File: xcp_d/utils/concantenation.py [hunk 6]
```diff
@@ -394,8 +414,24 @@ def concatenate_cifti(subid, fmridir, outputdir, ses=None, work_dir=None):
...
+            # print(resid)
+            for j in datafile:
+                fileid = res.split('task-')[0] + resid.partition('_desc')[0]
+                outfile = fileid + j
+                if j.endswith('framewisedisplacement_bold.tsv'):
+                    fileid = fileid.split('_den-91k')[0]
+                    filex = glob.glob(res.split('task-')[0] + '*task*' + j)
+                    name = f"{fileid}{j.split('.')[0]}-DCAN.hdf5"
+                    make_dcan_df(filex, name)
+        if len(resbold_unsmoothed_only) > 1:
```
Comment (tsalo) [inline xcp_d/utils/concantenation.py:434]:
```suggestion
                    make_dcan_df(filex, name)

        if len(resbold_unsmoothed_only) > 1:
```

File: xcp_d/utils/concantenation.py [hunk 7]
```diff
@@ -420,6 +456,9 @@ def concatenate_cifti(subid, fmridir, outputdir, ses=None, work_dir=None):
                     combine_fd(filex, outfile)
                     name = f"{fileid}{j.split('.')[0]}-DCAN.hdf5"
                     make_dcan_df(filex, name)
+                    for f in filex:
+                        name = f"{f.split('_space-')[0]}{j.split('.')[0]}-DCAN.hdf5" 
+                        make_dcan_df([f], name)
                 if j.endswith('dtseries.nii'):
```
Comment (tsalo) [inline xcp_d/utils/concantenation.py:462]:
```suggestion
                        make_dcan_df([f], name)

                if j.endswith('dtseries.nii'):
```

File: xcp_d/workflow/bold.py [hunk 8]
```diff
@@ -516,6 +546,7 @@ def init_boldpostprocess_wf(lower_bpf,
     workflow.connect([
         (filtering_wf, outputnode, [('filtered_file', 'processed_bold')]),
         (censor_scrub, outputnode, [('fd_timeseries', 'fd')]),
+        (censor_scrub, outputnode, [('tmask', 'tmask')]),
```
Comment (tsalo) [inline xcp_d/workflow/bold.py:549]: I think these two can be merged.

File: xcp_d/workflow/bold.py [hunk 9]
```diff
@@ -558,6 +589,8 @@ def init_boldpostprocess_wf(lower_bpf,
                                                   'inputnode.smoothed_bold')]),
         (censor_scrub, write_derivative_wf, [('fd_timeseries',
                                               'inputnode.fd')]),
+        (censor_scrub, write_derivative_wf, [('tmask',
+                                              'inputnode.tmask')]),
```
Comment (tsalo) [inline xcp_d/workflow/bold.py:593]: I think these two can be merged.

File: xcp_d/workflow/cifti.py [hunk 10]
```diff
@@ -453,6 +483,8 @@ def init_ciftipostprocess_wf(cifti_file,
                                                   'inputnode.smoothed_bold')]),
         (censor_scrub, write_derivative_wf, [('fd_timeseries',
                                               'inputnode.fd')]),
+        (censor_scrub, write_derivative_wf, [('tmask',
+                                              'inputnode.tmask')]),
```
Comment (tsalo) [inline xcp_d/workflow/cifti.py:487]: Can be combined.

File: xcp_d/workflow/cifti.py [hunk 11]
```diff
@@ -412,6 +441,7 @@ def init_ciftipostprocess_wf(cifti_file,
     workflow.connect([
         (filtering_wf, outputnode, [('filtered_file', 'processed_bold')]),
         (censor_scrub, outputnode, [('fd_timeseries', 'fd')]),
+        (censor_scrub, outputnode, [('tmask', 'tmask')]),
```
Comment (tsalo) [inline xcp_d/workflow/cifti.py:444]: Can be combined.

File: xcp_d/workflow/outputs.py [hunk 12]
```diff
@@ -448,6 +459,9 @@ def init_writederivatives_wf(
             (inputnode, write_derivative_gd333fc_wf, [('gd333_fc', 'in_file')]),
             (inputnode, write_derivative_ts50fc_wf, [('ts50_fc', 'in_file')]),
             (inputnode, write_derivative_fd_wf, [('fd', 'in_file')]),
+            (inputnode, write_derivative_tmask_wf, [('tmask', 'in_file')])
+
+
```
Comment (tsalo) [inline xcp_d/workflow/outputs.py:464]: I think these lines are extraneous.

File: xcp_d/workflow/postprocessing.py [hunk 13]
```diff
@@ -216,6 +216,65 @@ def fwhm2sigma(fwhm):
     return fwhm / np.sqrt(8 * np.log(2))
 
 
+def init_pre_smoothing(mem_gb,
+                       presmoothing,
+                       omp_nthreads,
+                       cifti=False,
+                       name="presmoothing"):
+    """Smooth input BOLD before postprocessing."""
```
Comment (tsalo) [inline xcp_d/workflow/postprocessing.py:224]: Can you document the parameters, inputs, and outputs?

File: xcp_d/workflow/postprocessing.py [hunk 14]
```diff
@@ -216,6 +216,65 @@ def fwhm2sigma(fwhm):
...
+                        name='inputnode')
+    outputnode = pe.Node(niu.IdentityInterface(fields=['presmoothed_bold']),
+                         name='outputnode')
+
+    sigma_lx = fwhm2sigma(presmoothing)  # Turn specified FWHM (Full-Width at Half Maximum)
+    # to standard deviation.
+    if cifti:  # For ciftis
+        workflow.__desc__ = f""" \
+The preprocessed BOLD input was pre-smoothed using Connectome Workbench with a
+gaussian kernel size of {str(presmoothing)} mm (FWHM).
```
Comment (tsalo) [inline xcp_d/workflow/postprocessing.py:236]:
```suggestion
Gaussian kernel size of {presmoothing} mm (FWHM).
```

I don't think you need to convert this to a string. The f-string formatter should handle it automatically.

File: xcp_d/workflow/postprocessing.py [hunk 15]
```diff
@@ -216,6 +216,65 @@ def fwhm2sigma(fwhm):
     return fwhm / np.sqrt(8 * np.log(2))
 
 
+def init_pre_smoothing(mem_gb,
```
Comment (tsalo) [inline xcp_d/workflow/postprocessing.py:219]: Other than the description, is this workflow any different from `init_resd_smoothing`? If not, maybe `init_resd_smoothing` could be generalized into `init_smoothing`, and the description could be either overridden outside the workflow or toggled based on a new parameter within the workflow.

**Review by tsalo — COMMENTED** [review]
I have a few formatting notes, but other than that it LGTM. I'm not sure if the new workflow is necessary though. Maybe you could just use a single smoothing workflow that is initialized twice for the two smoothing steps?

**Pull request closed without merging**